lib/Wikia/src/Service/User/Attributes/UserAttributes.php

```diff
@@ -67,55 +90,53 @@ public function setAttribute( $userId, Attribute $attribute ) {
 			return;
 		}
 
-		$this->setAttributeInService( $userId, $attribute );
-		$this->setAttributeInCache( $userId, $attribute );
-	}
-
-	private function isAnonUser( $userId ) {
-		return $userId === 0;
+		$this->setInService( $userId, $attribute );
+		$this->setInInstanceCache( $userId, $attribute );
+		$this->setInMemcache( $userId, $this->attributes[$userId] );
 	}
 
 	/**
 	 * @param $userId
-	 * @param array $attributes
+	 * @param Attribute $attribute
 	 */
-	public function setAttributesInCache( $userId, array $attributes ) {
-		foreach ( $attributes as $attributeKey => $attributeValue ) {
-			$this->setAttributeInCache( $userId, new Attribute( $attributeKey, $attributeValue ) );
-		}
+	private function setInService( $userId, $attribute ) {
+		$this->attributeService->set( $userId, $attribute );
 	}
 
 	/**
 	 * @param $userId
 	 * @param Attribute $attribute
 	 */
-	private function setAttributeInCache( $userId, Attribute $attribute ) {
+	private function setInInstanceCache( $userId, Attribute $attribute ) {
 		$this->attributes[$userId][$attribute->getName()] = $attribute->getValue();
 	}
 
-	/**
-	 * @param $userId
-	 * @param Attribute $attribute
-	 */
-	private function setAttributeInService( $userId, $attribute ) {
-		$this->attributeService->set( $userId, $attribute );
-	}
 
 	public function deleteAttribute( $userId, Attribute $attribute ) {
 		if ( $this->isAnonUser( $userId ) ) {
 			return;
 		}
 
-		$this->deleteAttributeFromService( $userId, $attribute );
-		$this->deleteAttributeFromCache( $userId, $attribute );
+		$this->deleteFromService( $userId, $attribute );
+		$this->deleteFromInstanceCache( $userId, $attribute );
+		$this->setInMemcache( $userId, $this->attributes[$userId] );
 	}
 
-	private function deleteAttributeFromCache( $userId, Attribute $attribute ) {
+	private function deleteFromService( $userId, Attribute $attribute ) {
+		$this->attributeService->delete( $userId, $attribute );
+	}
+
+	private function deleteFromInstanceCache( $userId, Attribute $attribute ) {
 		unset( $this->attributes[$userId][$attribute->getName()] );
 	}
 
-	private function deleteAttributeFromService( $userId, Attribute $attribute ) {
-		$this->attributeService->delete( $userId, $attribute );
+	private function isAnonUser( $userId ) {
```

Comment:
I think you're right. I was being a little paranoid about making sure anon users don't use the service, but it looks like `User.php` [checks for that](https://github.com/Wikia/app/blob/dev/includes/User.php#L3525) when it calls `saveSettings()`